File: backend/src/modules/live-learning/live-learning.controller.ts
import { Controller, Get, Post, Body, Param, HttpCode, HttpStatus, UseGuards, Req, ForbiddenException } from '@nestjs/common';
import { JwtAuthGuard } from '../auth/guards/jwt-auth.guard';
import { LiveLearningService, RoomLevel } from './live-learning.service';

@Controller('live')
@UseGuards(JwtAuthGuard)
export class LiveLearningController {
  constructor(private readonly liveService: LiveLearningService) {}

  /** Resolve acting user from token, rejecting client-supplied userId mismatches. */
  private self(sub: string, provided?: string): string {
    if (provided && provided !== sub) {
      throw new ForbiddenException('Cannot operate on another user');
    }
    return sub;
  }

  // Rooms
  @Get('rooms')
  getRooms() {
    const rooms = this.liveService.getRooms();
    return { success: true, data: rooms };
  }

  @Get('rooms/:id')
  getRoom(@Param('id') id: string) {
    const room = this.liveService.getRoom(id);
    return { success: true, data: room };
  }

  @Post('rooms/create')
  @HttpCode(HttpStatus.CREATED)
  createRoom(@Req() req, @Body() body: {
    hostId?: string;
    hostName: string;
    title: string;
    topic: string;
    description: string;
    level: RoomLevel;
    maxParticipants?: number;
    durationMinutes?: number;
    tags?: string[];
  }) {
    const room = this.liveService.createRoom({
      hostId: this.self(req.user.sub, body.hostId),
      hostName: body.hostName,
      title: body.title,
      topic: body.topic,
      description: body.description,
      level: body.level,
      maxParticipants: body.maxParticipants,
      durationMinutes: body.durationMinutes,
      tags: body.tags,
    });
    return { success: true, data: room };
  }

  @Post('rooms/:id/join')
  @HttpCode(HttpStatus.OK)
  joinRoom(@Req() req, @Param('id') id: string, @Body() body: { userId?: string; userName: string }) {
    const room = this.liveService.joinRoom(id, this.self(req.user.sub, body.userId), body.userName);
    return { success: true, data: room };
  }

  @Post('rooms/:id/leave')
  @HttpCode(HttpStatus.OK)
  leaveRoom(@Req() req, @Param('id') id: string, @Body() body: { userId?: string }) {
    const room = this.liveService.leaveRoom(id, this.self(req.user.sub, body.userId));
    return { success: true, data: room };
  }

  // Partners
  @Get('partners')
  getPartners() {
    const partners = this.liveService.getPartners();
    return { success: true, data: partners };
  }

  @Post('partners/match')
  @HttpCode(HttpStatus.OK)
  findMatch(@Req() req, @Body() body: {
    userId?: string;
    nativeLanguage: string;
    learningLanguage: string;
    level: string;
    goal: string;
  }) {
    const match = this.liveService.findMatch({
      userId: this.self(req.user.sub, body.userId),
      nativeLanguage: body.nativeLanguage,
      learningLanguage: body.learningLanguage,
      level: body.level,
      goal: body.goal,
    });
    return {
      success: true,
      data: match,
      message: match ? 'Match found!' : 'No match available. Try again later.',
    };
  }

  // Groups
  @Get('groups')
  getGroups() {
    const groups = this.liveService.getGroups();
    return { success: true, data: groups };
  }

  @Post('groups/:id/join')
  @HttpCode(HttpStatus.OK)
  joinGroup(@Req() req, @Param('id') id: string, @Body() body: { userId?: string; userName: string }) {
    const group = this.liveService.joinGroup(id, this.self(req.user.sub, body.userId), body.userName);
    return { success: true, data: group };
  }

  // Events
  @Get('events')
  getEvents() {
    const events = this.liveService.getEvents();
    return { success: true, data: events };
  }

  @Post('events/:id/join')
  @HttpCode(HttpStatus.OK)
  joinEvent(@Req() req, @Param('id') id: string, @Body() body: { userId?: string; userName: string }) {
    const event = this.liveService.joinEvent(id, this.self(req.user.sub, body.userId), body.userName);
    return { success: true, data: event };
  }

  // Analytics
  @Get('analytics')
  getAnalytics() {
    const analytics = this.liveService.getAnalytics();
    return { success: true, data: analytics };
  }
}

File: lib/features/live_learning/live_learning_controller.dart
import 'package:flutter/material.dart';
import 'package:lexi/core/services/auth_service.dart';
import 'package:lexi/features/live_learning/models/live_room.dart';
import 'package:lexi/features/live_learning/models/language_partner.dart';
import 'package:lexi/features/live_learning/models/learning_group.dart';
import 'package:lexi/features/live_learning/models/community_event.dart';
import 'package:lexi/features/live_learning/models/voice_message.dart';
import 'services/live_learning_service.dart';

class LiveLearningController extends ChangeNotifier {
  final LiveLearningService _service = LiveLearningService();
  List<LiveRoom> _rooms = [];
  List<LanguagePartner> _partners = [];
  List<LearningGroup> _groups = [];
  List<CommunityEvent> _events = [];
  List<VoiceMessage> _voiceMessages = [];
  LiveRoom? _currentRoom;
  bool _isLoading = false;
  String _error = '';
  List<LiveRoom> get rooms => _rooms;
  List<LanguagePartner> get partners => _partners;
  List<LearningGroup> get groups => _groups;
  List<CommunityEvent> get events => _events;
  List<VoiceMessage> get voiceMessages => _voiceMessages;
  LiveRoom? get currentRoom => _currentRoom;
  bool get isLoading => _isLoading;
  String get error => _error;
  Future<void> initialize() async {
    _isLoading = true;
    notifyListeners();
    try {
      _rooms = await _service.getRooms();
      _partners = await _service.getPartners();
      _groups = await _service.getGroups();
      _events = await _service.getEvents();
      _voiceMessages = await _service.getVoiceMessages(AuthService.instance.currentUser?.id ?? '');
    } catch (e) {
      _error = e.toString();
    }
    _isLoading = false;
    notifyListeners();
  }

  Future<LiveRoom> createRoom({
    required String hostId,
    required String hostName,
    required String title,
    required String topic,
    required String description,
    required RoomLevel level,
    int maxParticipants = 10,
    int durationMinutes = 30,
    List<String> tags = const [],
  }) async {
    _isLoading = true;
    notifyListeners();
    try {
      final room = await _service.createRoom(
        hostId: hostId,
        hostName: hostName,
        title: title,
        topic: topic,
        description: description,
        level: level,
        maxParticipants: maxParticipants,
        durationMinutes: durationMinutes,
        tags: tags,
      );
      _rooms.insert(0, room);
      _isLoading = false;
      notifyListeners();
      return room;
    } catch (e) {
      debugPrint('Error creating room: $e');
      _isLoading = false;
      notifyListeners();
      rethrow;
    }
  }

  Future<void> joinRoom(String roomId, String userId, String userName) async {
    _isLoading = true;
    notifyListeners();
    try {
      final room = await _service.joinRoom(roomId, userId, userName);
      if (room != null) {
        final index = _rooms.indexWhere((r) => r.id == roomId);
        if (index != -1) {
          _rooms[index] = room;
        }
        _currentRoom = room;
      }
    } catch (e) {
      debugPrint('Error joining room: $e');
      _error = e.toString();
    }
    _isLoading = false;
    notifyListeners();
  }

  Future<void> leaveRoom(String roomId, String userId) async {
    _isLoading = true;
    notifyListeners();
    try {
      final room = await _service.leaveRoom(roomId, userId);
      if (room != null) {
        final index = _rooms.indexWhere((r) => r.id == roomId);
        if (index != -1) {
          _rooms[index] = room;
        }
        if (_currentRoom?.id == roomId) {
          _currentRoom = null;
        }
      }
    } catch (e) {
      debugPrint('Error leaving room: $e');
    }
    _isLoading = false;
    notifyListeners();
  }

  Future<PartnerMatch?> findPartnerMatch({
    required String userId,
    required String nativeLanguage,
    required String learningLanguage,
    required String level,
    required MatchingGoal goal,
  }) async {
    _isLoading = true;
    notifyListeners();
    try {
      final match = await _service.findMatch(
        userId: userId,
        nativeLanguage: nativeLanguage,
        learningLanguage: learningLanguage,
        level: level,
        goal: goal,
      );
      _isLoading = false;
      notifyListeners();
      return match;
    } catch (e) {
      debugPrint('Error finding partner match: $e');
      _isLoading = false;
      notifyListeners();
      return null;
    }
  }

  Future<void> joinGroup(String groupId, String userId, String userName) async {
    _isLoading = true;
    notifyListeners();
    try {
      final group = await _service.joinGroup(groupId, userId, userName);
      if (group != null) {
        final index = _groups.indexWhere((g) => g.id == groupId);
        if (index != -1) {
          _groups[index] = group;
        }
      }
    } catch (e) {
      debugPrint('Error joining group: $e');
    }
    _isLoading = false;
    notifyListeners();
  }

  Future<void> joinEvent(String eventId, String userId, String userName) async {
    _isLoading = true;
    notifyListeners();
    try {
      final event = await _service.joinEvent(eventId, userId, userName);
      if (event != null) {
        final index = _events.indexWhere((e) => e.id == eventId);
        if (index != -1) {
          _events[index] = event;
        }
      }
    } catch (e) {
      debugPrint('Error joining event: $e');
    }
    _isLoading = false;
    notifyListeners();
  }

  Future<void> sendVoiceMessage(VoiceMessage message) async {
    await _service.saveVoiceMessage(message);
    _voiceMessages.insert(0, message);
    notifyListeners();
  }

  void setCurrentRoom(LiveRoom? room) {
    _currentRoom = room;
    notifyListeners();
  }

  void clearError() {
    _error = '';
    notifyListeners();
  }
}
